ddtrace/internal/_encoding.pyx

```diff
@@ -0,0 +1,357 @@
+                if L > ITEM_LIMIT:
+                    raise ValueError("dict is too large")
+                ret = msgpack_pack_map(&self.pk, L)
+                if ret == 0:
+                    for k, v in d.items():
+                       ret = self._pack(k)
+                       if ret != 0: break
+                       ret = self._pack(v)
+                       if ret != 0: break
+            elif PyList_CheckExact(o):
+                # Expect a list of traces or a list of spans
+
+                L = len(o)
+                if L > ITEM_LIMIT:
+                    raise ValueError("list is too large")
+
+                ret = msgpack_pack_array(&self.pk, L)
+                if ret != 0:
+                    break
+
+                if L > 0 and PyList_CheckExact(o[0]):
+                    # List of traces
+                    for i in range(L):
+                        span = o[i]
+                        ret = self._pack(span)
+                        if ret != 0: break
+                else:
+                    # List of spans
+                    for i in range(L):
+                        ret = self._pack(o[i])
+                        if ret != 0: break
+
+            elif type(o) is Span:
+                L = 12
+                if L > ITEM_LIMIT:
+                    raise ValueError("list is too large")
+
+                has_span_type = <bint>(o.span_type is not None)
+                has_meta = <bint>(len(o.meta) > 0)
+                has_metrics = <bint>(len(o.metrics) > 0)
+
+                L = L - (1 - has_span_type) - (1 - has_meta) - (1 - has_metrics)
+
+                ret = msgpack_pack_map(&self.pk, L)
+
+                if ret == 0:
+                    ret = self._pack_bytes(<char *>b"trace_id")
+                    if ret != 0: return ret
+                    ret = self._pack(o.trace_id)
+                    if ret != 0: return ret
+
+                    ret = self._pack_bytes(<char *>b"parent_id")
+                    if ret != 0: return ret
+                    ret = self._pack(o.parent_id)
+                    if ret != 0: return ret
+
+                    ret = self._pack_bytes(<char *>b"span_id")
+                    if ret != 0: return ret
+                    ret = self._pack(o.span_id)
+                    if ret != 0: return ret
+
+                    ret = self._pack_bytes(<char *>b"service")
+                    if ret != 0: return ret
+                    ret = self._pack(o.service)
+                    if ret != 0: return ret
+
+                    ret = self._pack_bytes(<char *>b"resource")
+                    if ret != 0: return ret
+                    ret = self._pack(o.resource)
+                    if ret != 0: return ret
+
+                    ret = self._pack_bytes(<char *>b"name")
+                    if ret != 0: return ret
+                    ret = self._pack(o.name)
+                    if ret != 0: return ret
+
+                    ret = self._pack_bytes(<char *>b"error")
```

Comment:
I wonder if we can just manually append bytes into the buffer, since this is basically:

```python
self.buffer += b'\xa5error\x0{}'.format(1 if o.error else 0)
```

Reply:
Tricky part is the bounds checking to make sure we don't overrun / allocate the next chunk of memory.

Reply:
but in theory for ones like this the length is static so:

```python
# bounds check on 9 bytes
# append b'\xa5error\x0{}'.format(1 if o.error else 0)
```

which is probably mostly what `self._pack_bytes` is doing, but if we can move packing the key and the value together into one call, we can gain something from that